### PR TITLE
Iterated type families

### PR DESCRIPTION
**Pull request by EgbertRijke**
This is an experimental attempt to get some sort of uniform treatment of iterated type families going.

Work in progress, and ideas are very welcome.

### CODE REVIEW

File: src/foundation/iterated-type-families.lagda.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+## Idea
+
+An **iterated type family** is a sequence of type families
+`(A₀, A₁, A₂, ..., A_n)` consisting of
+
+- a type `A₀`,
+- a type family `A₁ : A₀ → 𝒰`,
+- a type family `A₂ : (x₀ : A₀) → A₁ x₀ → 𝒰`,
+- ...
+- a type family `An : (x0 : A₀) ... (x*(n-1) : A*(n-1) x₀ ... x\_(n-2)) → 𝒰.
```
Comment (fredrik-bakke) [inline src/foundation/iterated-type-families.lagda.md:30]:
```suggestion
- a type family `An : (x0 : A₀) ... (x*(n-1) : A*(n-1) x₀ ... x\_(n-2)) → 𝒰`.
```

File: src/trees/universal-tree.lagda.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ensure that it contains a tree of
+[iterated type families](foundation.iterated-type-families.md) of which each
+type family `A_i` is assigned its own universe level.
+
+```agda
+record Universal-Large-Tree : UUω
+  where
+  coinductive
+  field
+    level-Universal-Large-Tree : Level
```
Comment (fredrik-bakke) [inline src/trees/universal-tree.lagda.md:64]: Do you need the level as a field rather than a parameter?

Comment (EgbertRijke) [reply src/trees/universal-tree.lagda.md:64]: I realize that the explanatory paragraph mentions levels as parameters. That's my mistake. In this case I consciously chose it to be a field, because that way the levels can be different for each family in the list of iterated families.

Comment (fredrik-bakke) [reply src/trees/universal-tree.lagda.md:64]: Right, sorry for the stupid question! Coinductive types isn't my forte 😅

Comment (EgbertRijke) [reply src/trees/universal-tree.lagda.md:64]: It was a very valid question.

Coinductive types are terminal coalgebras of polynomial endofunctors. Those can all be constructed as squential limits. But I'm not 100% sure if I will need the full power of coinductive types. Maybe that's just what led me to the idea... We'll see.

**Comment by fredrik-bakke** [comment]
Okay, that is pretty damn cool

**Comment by fredrik-bakke** [comment]
Now I have to start compiling a list of theorems we can use this for!

**Comment by EgbertRijke** [comment]
Yup, I got something working that computes definitionally to a general iterated dependent product. Now the hope is that these iterated type families can be implicit arguments in theorems like `is-prop-iterated-Pi`. But I go to sleep now.

**Comment by EgbertRijke** [comment]
I'll be very happy to have your list of theorems.

**Comment by fredrik-bakke** [comment]
Sleep tight! We'll speak later

File: src/foundation/iterated-type-families.lagda.md [hunk 3]
```diff
@@ -49,4 +53,43 @@ tree.
...
+  {A : UU l1} {B : A → UU l2} {C : (x : A) → B x → UU l3}
+  where
+
+  test :
+    Π-Iterated-Type-Family
+      ( cons-Iterated-Type-Family
+        ( λ x → cons-Iterated-Type-Family
+          ( λ y → base-Iterated-Type-Family (C x y)))) ＝
+    ( (x : A) (y : B x) → C x y)
+  test = refl
```
Comment (EgbertRijke) [inline src/foundation/iterated-type-families.lagda.md:94]: 👆😁

**Comment by fredrik-bakke** [comment]
I think it should pretty immediately give us iterated theorems for all constructions of Πs and Σs that are "uniform" in one variable. Like
- `is-prop-Π` `is-prop-implicit-Π` (soon to be renamed)
- `is-contr-total-Eq-Π` and `is-contr-total-Eq-Π`
- `equiv-Π-equiv-family`
- `equiv-tot`
Moreover, we should be able to make variant theorems that are indexed in binary trees, so we can index a point in a type composite Sigma or Pi type where we want to insert an equivalence.

**Comment by fredrik-bakke** [comment]
I'll have to play a little around with it, but can it also give us an iterated `is-contr-Σ`? That would be interesting to see.

**Comment by EgbertRijke** [comment]
> I'll have to play a little around with it, but can it also give us an iterated `is-contr-Σ`? That would be interesting to see.

This would indeed be nice. Given that we want to be able to do this, we need to be able to speak of iterated familes _over_ iterated families. That reminds me that the correct terminology for iterated family is _telescope_. Using this terminology, it is also easier to speak of dependent telescopes. So I'm going to rename `iterated-type-family` to `telescope`. 

Once I have the notion of dependent telescopes I think we can also attempt to iterate the structure identity principle.

**Comment by fredrik-bakke** [comment]
Are we also able to speak of any composite of just Σ or Π types by replacing the natural number parameter with a standard binary tree? I wonder if this will allow us to define a general `associative-Σ` that can just infer the endpoints. If that is the case, can we make the type checker also infer a general composite of Σs _and_ Πs?

**Comment by EgbertRijke** [comment]
That sounds more complicated. Let’s first try to accomplish our original goals.

**Comment by fredrik-bakke** [comment]
Sure! I'm just spitting ideas

**Comment by fredrik-bakke** [comment]
Another "simpler" one is iterated `universally-structured-Π`

File: src/foundation/homotopies-iterated-dependent-product-types.lagda.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+## Definitions
+
+### Homotopies in iterated dependent products of iterated type families
+
+```agda
+iterated-htpy :
+  {l : Level} {n : ℕ} {{A : telescope l n}} (f g : iterated-Π A) → UU l
+iterated-htpy {{base-telescope A}} f g = f ＝ g
+iterated-htpy {{cons-telescope A}} f g =
+  (x : _) → iterated-htpy {{A x}} (f x) (g x)
```
Comment (EgbertRijke) [inline src/foundation/homotopies-iterated-dependent-product-types.lagda.md:47]: Should this not be defined as an iterated Pi-type?

Comment (fredrik-bakke) [reply src/foundation/homotopies-iterated-dependent-product-types.lagda.md:47]: If you have a particular nice definition in mind then be my guest!

Comment (fredrik-bakke) [reply src/foundation/homotopies-iterated-dependent-product-types.lagda.md:47]: By the way, after a little more consideration, I think something like `multivariable-homotopies` may be a better name for this file. This makes more sense when defining things like `multivariable-sections` and `multivariable-equivalences`

Comment (EgbertRijke) [reply src/foundation/homotopies-iterated-dependent-product-types.lagda.md:47]: I agree, that's a great name

**Comment by fredrik-bakke** [comment]
I just noticed that we have the following definitions in our library:
- `apply-twice-universal-property-trunc-Prop`
- `apply-three-times-universal-property-trunc-Prop`
- `apply-twice-dependent-universal-property-surj-is-surjective`

It seems fitting to replace these with general iterated versions after this PR is merged, so I'm recording it here :)

File: src/trees/universal-tree.lagda.md (hunk 2 above)
Comment (fredrik-bakke) [reply src/trees/universal-tree.lagda.md:63]: I've mentioned it elsewhere, but I'll write it again here so it isn't forgotten. There is the `--no-level-universe` flag that can be set for files. So if we enable `--level-universe` globally for the library, we can still disable it locally in this file and its derivatives by setting that flag.

Comment (EgbertRijke) [reply src/trees/universal-tree.lagda.md:63]: Good call! Indeed I had already forgotten about it:)

Comment (fredrik-bakke) [reply src/trees/universal-tree.lagda.md:63]: Lol! I figured you might have 😄

**Comment by EgbertRijke** [comment]
> I just noticed that we have the following definitions in our library:
> 
> * `apply-twice-universal-property-trunc-Prop`
> * `apply-three-times-universal-property-trunc-Prop`
> * `apply-twice-dependent-universal-property-surj-is-surjective`
> 
> It seems fitting to replace these with general iterated versions after this PR is merged, so I'm recording it here :)

We also have some `double-induction` entries

**Comment by EgbertRijke** [comment]
I compiles, therefore it is ready for review 🤷

**Comment by fredrik-bakke** [comment]
I'll do it once I finish my own PR which will be like 10-20 minutes! 🚀🚀🚀

File: src/foundation/homotopies-iterated-dependent-product-types.lagda.md (hunk 4 above)
Comment (EgbertRijke) [reply src/foundation/homotopies-iterated-dependent-product-types.lagda.md]: This comment is now addressed. Great suggestion!

File: src/trees/universal-tree.lagda.md (hunk 2 above)
Comment (EgbertRijke) [reply src/trees/universal-tree.lagda.md:63]: I just remove the universal large tree for now. I have no plans for it anymore anyway.

**Comment by EgbertRijke** [comment]
But but... there are no more unresolved converstations... 

Hit that merge button Fredrik! 😄

**Comment by fredrik-bakke** [comment]
Can I please have a quick review first?

**Comment by EgbertRijke** [comment]
I think you wrote most if this PR actually

**Comment by fredrik-bakke** [comment]
Links 👀

File: agda-unimath.agda-lib [hunk 5]
```diff
@@ -1,3 +1,3 @@
 name: agda-unimath
 include: src
-flags: --without-K --exact-split --no-import-sorts --auto-inline
```
Comment (fredrik-bakke) [inline agda-unimath.agda-lib:3]: Is this right that we want to assume it globally?

Comment (EgbertRijke) [reply agda-unimath.agda-lib:3]: No... that's a mistake. I don't know how it got there. Thanks for catching it!

File: src/elementary-number-theory/natural-numbers.lagda.md [hunk 6]
```diff
@@ -121,7 +120,7 @@ is-successor-is-nonzero-ℕ {zero-ℕ} H = ex-falso (H refl)
 pr1 (is-successor-is-nonzero-ℕ {succ-ℕ x} H) = x
 pr2 (is-successor-is-nonzero-ℕ {succ-ℕ x} H) = refl
 
-has-no-fixed-points-succ-ℕ : (x : ℕ) → succ-ℕ x ≠ x
+has-no-fixed-points-succ-ℕ : (x : ℕ) → ¬ (succ-ℕ x ＝ x)
 has-no-fixed-points-succ-ℕ x ()
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/natural-numbers.lagda.md:124]: Maybe just revert the changes to this file and merge #876 before this one.

File: src/foundation/dependent-telescopes.lagda.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+open import foundation.telescopes
+open import foundation.universe-levels
+```
+
+</details>
+
+## Idea
+
+A **dependent telescope** over a [telescope](foundation.telescopes.md) `A` of
+length `n` is a dependent list of dependent types over each of the entries in
```
Comment (fredrik-bakke) [inline src/foundation/dependent-telescopes.lagda.md:23]: here you use "length", elsewhere, "depth" is used. Length sounds more natural to me, but you can choose.

File: src/foundation/dependent-telescopes.lagda.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  cons-dependent-telescope :
+    {l1 l2 k1 k2 : Level} {n : ℕ} {X : UU l1} {A : X → telescope l2 n}
+    {Y : X → UU k1} (B : (x : X) → Y x → dependent-telescope k2 (A x)) →
+    dependent-telescope (k1 ⊔ k2) (cons-telescope A)
+```
+
+### Expansion of telescopes
+
+An **expansion** of a telescope `A` by a dependent telescope `B` over it is a
+new telescope of the same depth as `A`, constructed by taking
```
Comment (fredrik-bakke) [inline src/foundation/dependent-telescopes.lagda.md:61]: depth/length

File: src/foundation/dependent-telescopes.lagda.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+expand-telescope (base-dependent-telescope Y) =
+  base-telescope (Σ _ Y)
+expand-telescope (cons-dependent-telescope B) =
+  cons-telescope (λ x → expand-telescope (B (pr1 x) (pr2 x)))
+```
+
+### Interleaving telescopes
+
+Given a telescope `A` of depth `n` and a dependent telescope `B` over it, we can
+define the **interleaved telescope** whose depth is `2n`.
```
Comment (fredrik-bakke) [inline src/foundation/dependent-telescopes.lagda.md:77]: length/depth

File: src/foundation/dependent-telescopes.lagda.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+```agda
+telescope-Π :
+  {l1 l2 : Level} {n : ℕ} {A : telescope l1 n} →
+  dependent-telescope l2 A → telescope (l1 ⊔ l2) n
+telescope-Π (base-dependent-telescope Y) =
+  base-telescope ((x : _) → Y x)
+telescope-Π (cons-dependent-telescope B) =
+  cons-telescope (λ x → telescope-Π (B (pr1 x) (pr2 x)))
+```
```
Comment (fredrik-bakke) [inline src/foundation/dependent-telescopes.lagda.md:102]: "mapping telescope" is just made-up terminology. We don't have to change it now, but maybe it should be reconsidered in the future

File: src/foundation/homotopies.lagda.md [hunk 11]
```diff
@@ -311,6 +311,6 @@ module _
 
 ## See also
 
-- We postulate that homotopies characterize identifications in (dependent)
-  function types in the file
-  [`foundation-core.function-extensionality`](foundation-core.function-extensionality.md).
```
Comment (fredrik-bakke) [inline src/foundation/homotopies.lagda.md:316]: Oops, looks like this was overwritten by mistake

File: src/foundation/dependent-telescopes.lagda.md (hunk 10 above)
Comment (EgbertRijke) [reply src/foundation/dependent-telescopes.lagda.md:102]: Sure!

File: src/foundation/telescopes.lagda.md [hunk 12]
```diff
@@ -0,0 +1,875 @@
...
+
+- a type `A₀`,
+- a type family `A₁ : A₀ → 𝒰`,
+- a type family `A₂ : (x₀ : A₀) → A₁ x₀ → 𝒰`,
+- ...
+- a type family `A_n : (x₀ : A₀) ... (x_(n-1) : A_(n-1) x₀ ... x_(n-2)) → 𝒰`.
+
+We say that a telescope `(A₀,...,A_n)` has **length** `n+1`. In other words, the
+length of the telescope `(A₀,...,A_n)` is the length of the (dependent) list
+`(A₀,...,A_n)`.
```
Comment (fredrik-bakke) [inline src/foundation/telescopes.lagda.md:30]: length/depth

File: src/foundation/telescopes.lagda.md [hunk 13]
```diff
@@ -0,0 +1,875 @@
...
+The type of telescopes is a [directed tree](trees.directed-trees.md)
+
+```text
+  ... → T₃ → T₂ → T₁ → T₀,
+```
+
+where `T_n` is the type of all telescopes of length `n`, and the map from
+`T_(n+1)` to `T_n` maps `(A₀,...,A_n)` to `(A₀,...,A_(n-1))`. The type of such
+directed trees can be defined as a coinductive record type, and we will define
+the tree `T` of telescopes as a particular element of this tree.
```
Comment (fredrik-bakke) [inline src/foundation/telescopes.lagda.md:47]: The relation of telescopes to this type of trees is not made formal.

File: src/foundation/telescopes.lagda.md [hunk 14]
```diff
@@ -0,0 +1,873 @@
...
+```agda
+apply-base-telescope :
+  {l1 : Level} {n : ℕ}
+  (P : {l : Level} → UU l → UU l) → telescope l1 n → telescope l1 n
+apply-base-telescope P (base-telescope A) = base-telescope (P A)
+apply-base-telescope P (cons-telescope A) =
+  cons-telescope (λ x → apply-base-telescope P (A x))
+```
+
+### Inferring telescopes
```
Comment (fredrik-bakke) [inline src/foundation/telescopes.lagda.md:79]:
```suggestion
### Telescopes as instance arguments
```

File: src/foundation/telescopes.lagda.md [hunk 15]
```diff
@@ -0,0 +1,873 @@
...
+
+To get Agda to infer telescopes, we help it along a little using
+[instance arguments](https://agda.readthedocs.io/en/v2.6.3.20230914/language/instance-arguments.html).
```
Comment (fredrik-bakke) [inline src/foundation/telescopes.lagda.md:82]:
```suggestion
[instance arguments](https://agda.readthedocs.io/en/latest/language/instance-arguments.html).
```

File: src/foundation/telescopes.lagda.md [hunk 16]
```diff
@@ -0,0 +1,873 @@
...
+These are a special kind of implicit argument in Agda that are resolved by the
```
Comment (fredrik-bakke) [inline src/foundation/telescopes.lagda.md:83]:
```suggestion
These is a special kind of implicit argument in Agda that are resolved by the
```

File: src/foundation/homotopies.lagda.md (hunk 11 above)
Comment (EgbertRijke) [reply src/foundation/homotopies.lagda.md:316]: No I removed it, since this file already uses function extensionality. The homotopies file in foundation-core still mentions it.

Is that not ok?

File: src/foundation/telescopes.lagda.md (hunk 13 above)
Comment (EgbertRijke) [reply src/foundation/telescopes.lagda.md:47]: Lol, here you see your towers:)

Comment (EgbertRijke) [reply src/foundation/telescopes.lagda.md:47]: Should this text mention that it is not made formal, or is it clear from the context (that it is in the idea section) that this is a sketch of the idea?

File: src/foundation/telescopes.lagda.md [hunk 17]
```diff
@@ -0,0 +1,873 @@
...
+instance resolution algorithm. We register building blocks for this algorithm to
+use below, i.e. _instances_. Then Agda will attempt to use those to construct
+telescopes of the appropriate kind when asked to.
+
```
Comment (fredrik-bakke) [inline src/foundation/telescopes.lagda.md:87]:
```suggestion

In the case of telescopes, this has the unfortunate disadvantage that we can only define instances for fixed length telescopes. We have defined instances of telescopes up to length 18, so although Agda cannot infer a telescope of a general length using this approach, it can infer them up to this given length.

```

File: src/elementary-number-theory/natural-numbers.lagda.md (hunk 6 above)
Comment (EgbertRijke) [reply src/elementary-number-theory/natural-numbers.lagda.md]: Already did

**Comment by EgbertRijke** [comment]
> Links 👀

Links where?

**Comment by fredrik-bakke** [comment]
> > Links 👀
> 
> Links where?

the link check failed

File: src/trees/empty-multisets.lagda.md [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+```agda
+module _
+  {l : Level}
+  where
+
+  is-empty-𝕍 : 𝕍 l → UU l
+  is-empty-𝕍 (tree-𝕎 X Y) = is-empty X
+```
+
+### The pradicate of being a multiset with no elements
```
Comment (fredrik-bakke) [inline src/trees/empty-multisets.lagda.md:40]:
```suggestion
### The predicate of being a multiset with no elements
```

**Comment by EgbertRijke** [comment]
On my end everything is compiling now

File: src/foundation/telescopes.lagda.md (hunk 13 above)
Comment (fredrik-bakke) [reply src/foundation/telescopes.lagda.md:47]: Oh, I thought it was saying something about how telescopes are defined, but it is not. Never mind!

File: src/trees/multiset-indexed-dependent-products-of-types.lagda.md [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+open import trees.w-types
+```
+
+</details>
+
+## Idea
+
+Consider a [multiset](trees.multisets.md) `M`. Then `M` can be seen as a tower
+of type families, via the inclusion from the type of all multisets, which are
+the well-founded trees, into the type of all trees.
```
Comment (fredrik-bakke) [inline src/trees/multiset-indexed-dependent-products-of-types.lagda.md:25]: are these defined anywhere? if so, links :)

Comment (fredrik-bakke) [reply src/trees/multiset-indexed-dependent-products-of-types.lagda.md:25]: well-founded trees and trees that is

File: src/foundation/telescopes.lagda.md (hunk 16 above)
Comment (EgbertRijke) [reply src/foundation/telescopes.lagda.md:83]: Lol why are you introducing a typo?

Comment (fredrik-bakke) [reply src/foundation/telescopes.lagda.md:83]: Gotta keep you on your toes

File: src/trees/multiset-indexed-dependent-products-of-types.lagda.md (hunk 19 above)
Comment (EgbertRijke) [reply src/trees/multiset-indexed-dependent-products-of-types.lagda.md:25]: of course! `trees` is full of that stuff

**Comment by fredrik-bakke** [comment]
alright, that's it, I'm glad to have this merged once the last comments are resolved :)

File: src/trees/multiset-indexed-dependent-products-of-types.lagda.md (hunk 19 above)
Comment (EgbertRijke) [reply src/trees/multiset-indexed-dependent-products-of-types.lagda.md:25]: W-types are well-founded by definition, and they're a bit more than trees. They're enriched trees.

Comment (fredrik-bakke) [reply src/trees/multiset-indexed-dependent-products-of-types.lagda.md:25]: Sorry, I'm too tired to apply critical thinking atm 😅

Comment (fredrik-bakke) [reply src/trees/multiset-indexed-dependent-products-of-types.lagda.md:25]: That was the one thought I _did_ have 😅

Comment (EgbertRijke) [reply src/trees/multiset-indexed-dependent-products-of-types.lagda.md:25]: I don't really know what to link to at the moment. Maybe I can come back to this and nicely link it up later.

File: src/foundation/homotopies.lagda.md (hunk 11 above)
Comment (fredrik-bakke) [reply src/foundation/homotopies.lagda.md:316]: gotcha, that's fine

**Comment by EgbertRijke** [comment]
Alright, I ran `make pre-commit` once more. It should be good now
